### ISSUE TITLE
Consider switching from zisofs to squashfs

### ISSUE THREAD

**Issue by probonopd** [probonopd/AppImageKit#234]
Consider switching from zisofs to squashfs. This would:
- Significantly save space, up to 45% (e.g., 121 MB rather than 218 MB for QtCreator-5.7.0-x86_64.AppImage when using a 1 MB block size with `-comp xz -b 1M` but that makes application launches of large apps noticeably slower); or 27% (159 MB rather than 218 MB) with great performance using default `gzip2` compression
- Allow us to use less arcane generation tools, i.e., mksquashfs instead of xorriso
- Significantly speed up AppImage generation time, around 40% (e.g., 1:38 instead of 2:37 for QtCreator-5.7.0-x86_64.AppImage)

We would lose:
- Mounting with Disk Image Mounter (until it supports offsets)
- Extracting with ark (until it supports squashfs with offsets)

To be decided:
- Where to store updateinformation. Proposal: ELF PT_NOTE section `.note.upd-info`.
- Use gzip2 or lz compression. Tradeoff performance vs. compression ratio. Might support both. proposal: gzip2 by default

Proof of concept:

**EDIT: Now at [https://github.com/probonopd/appimageruntime](https://github.com/probonopd/appimageruntime)**

```
git clone https://github.com/vasi/squashfuse.git
cd squashfuse/
sudo apt install -y autoconf libtool make gcc libtool libtool-bin libfuse-dev liblzma-dev
libtoolize --force
aclocal
autoheader
automake --force-missing --add-missing
autoconf

./configure --with-xz=/usr/lib/

make

mkdir mountpoint

mksquashfs /mnt/ xchat.sfs -root-owned -noappend

# Create 64k of padding and append squashfs to the padding
dd if=xchat.sfs of=xchat.sai bs=1024 seek=128

# Now let's embed the runtime
strip squashfuse
dd conv=notrunc if=squashfuse of=xchat.sai
chmod +x xchat.sai

# Inject Subtype 2 AppImage magic
printf '\x41\x49\x02' | dd of=xchat.sai bs=1 seek=8 count=3 conv=notrunc

# Sure enough it can mount itself
./xchat.sai -o offset=$((128*1024)) xchat.sai mountpoint/

sudo umount mountpoint/ # Why do I need sudo here?

# Can the kernel do it too? Sure enough
sudo mount -o loop,offset=$((128*1024)) xchat.sai mountpoint/

sudo umount mountpoint/ 
```

**Comment by PHPJosh** [comment]
should be able to mount with code like this.

`mount App.appimage /mnt/Appdir -t squashfs -o loop`

probably consider writing a helper script  that works with the Appimage mime-type as some folks like taking a peak inside much like people can see the Contents of macOS apps.

**Comment by probonopd** [comment]
Nooooo... the AppImage is supposed to mount __itself__

**Comment by PHPJosh** [comment]
yep i know was mainly talking about adding a open with  entry on context menus so people can manually mount if they want too. which would be real similar to macOS's "Show Contents" context entry as people should be able to peak inside easily if they want without hiting the command line if they want

**Comment by probonopd** [comment]
That could still be done by the AppImage itself, e.g., using a `--mount-only` option to be implemented.

**Comment by probonopd** [comment]
Note to self on making AppImageKit runtime use another fuse fs:

```
# In the fuse fs (e.g., squashfusefs ll.c), change main to fusefs_main

change in runtime.c the line starting with "extern int fusefs_main" to
extern int fusefs_main(int argc, char *argv[]);
because must match whatever is in the fuse fs

# Compile the fuse fs
# It will error out saying it is missing main but that is OK

# In runtime.c, change ext2_main to fusefs_main (2 locations!) and add the offset 65536 to the options

# Compile runtime but do not link
cc -D_FILE_OFFSET_BITS=64 -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -g -Os -c runtime.c

# Now link but against libsquashfuse
cc runtime.o ../squashfuse_ll-ll_inode.o ../squashfuse_ll-nonstd-daemon.o ../squashfuse_ll-ll.o -o runtime -rdynamic ../.libs/libsquashfuse.a ../.libs/libfuseprivate.a -lfuse -lpthread -lglib-2.0 -lz -llzma

strip runtime
```
Along these lines; not quite working yet since argv seems to to be passed to the fuse fs correctly.

**Comment by probonopd** [comment]
Could consider putting data into ELF sections
Pro:
* Easy to extract
* Do not need fixed offsets
Cons:
* Need objcopy tool to extract
* Need gcc to create the ELF with the sections each time an AppImage is packaged - might be a no-go (think e.g., electron-builder on macOS producing AppImages for Linux)

**Comment by probonopd** [comment]
We now change
```
cc runtime.o ./squashfuse/squashfuse_ll-ll_inode.o ./squashfuse/squashfuse_ll-nonstd-daemon.o ./squashfuse/squashfuse_ll-ll.o -o runtime  ./squashfuse/.libs/libsquashfuse.a ./squashfuse/.libs/libfuseprivate.a   -Wl,-Bdynamic -lfuse -lpthread -lglib-2.0 -lz -Wl,-Bstatic -llzma -Wl,-Bdynamic
```
so that the squashfuse_ll stuff becomes a library
```
ar rcs libsquashfusell.a ./squashfuse/squashfuse_ll-ll_inode.o ./squashfuse/squashfuse_ll-nonstd-daemon.o ./squashfuse/squashfuse_ll-ll.o
```
We can now link against that library instead of the individual -o files:
```
cc runtime.o -o runtime libsquashfusell.a ./squashfuse/.libs/libsquashfuse.a ./squashfuse/.libs/libfuseprivate.a -Wl,-Bdynamic -lfuse -lpthread -lglib-2.0 -lz -Wl,-Bstatic -llzma -Wl,-Bdynamic
```

**Comment by probonopd** [comment]
https://github.com/probonopd/appimageruntime

**Comment by probonopd** [comment]
https://github.com/AppImage/AppImageSpec/blob/master/draft.md#type-2-image-format

**Comment by hideout** [comment]
works

2016-09-18 22:47 GMT+02:00 probonopd <notifications@github.com>:

**Comment by bignaux** [comment]
squashfs in last file-roller ? https://launchpad.net/ubuntu/+source/file-roller/3.22.0-1ubuntu1
I've no better result with lz4 -Xhc in matter of compression : 196MB with -b1M and 201MB by default. I should try decompression according to https://fastcompression.blogspot.fr/2015/01/zstd-stronger-compression-algorithm.html should be far better with it.

**Comment by probonopd** [comment]
Looks like they patched it in: 

> debian/patches/squashfs.patch:
+ Support squashfs and snap files
+ Suggest p7zip-full and squashfs-tools

So they are probably using p7zip-full and/or squashfs-tools

https://launchpad.net/ubuntu/+source/file-roller/3.22.0-1ubuntu1

**Comment by probonopd** [comment]
Zstandard looks interesting, could squashfs use it as a compressor?

**Comment by bignaux** [comment]
For daily use where only decompression is important, LZ4 is 3.5x faster than zstd AND available. Who want to spend time to add zstd in all our requirement to gain a bit on compression time ...

**Comment by probonopd** [comment]
@bignaux agree, let's not worry about compression time. _Decompression_ time is what matters.

**Comment by probonopd** [comment]
Read http://www.nongnu.org/lzip/xz_inadequate.html on why xz/lzma is seen as inadequate for long-term archiving by the author.

**Comment by probonopd** [comment]
Switch complete.

**Comment by probonopd** [comment]
@bignaux looks like we can start using zstd, https://github.com/AppImage/AppImageKit/issues/478

**Comment by harrytuttle** [comment]
> Mounting with Disk Image Mounter (until it supports offsets)

Would it be possible to inject a dos partition table at bytes 446-512? The AppImage becomes mountable by gnome-disk-image-mounter, and by pure chance x86_64 elf binaries appears to be working as before. i386 binaries segfault unless there is a way to move the early elf data structures or something like that.

**Comment by probonopd** [comment]
Can you attach an example file @harrytuttle? Sounds interesting.

**Comment by TheAssassin** [comment]
Please open another issue about that (or start a PR directly, if you want to), please.

**Comment by harrytuttle** [comment]
Nevermind, I can't make it work anymore, i'll open an issue if i'll find an example.
Still, i'd recommend padding the runtime to a multiple of 512bytes, it will be easier to convert an appimage to a mountable disk image and back.

**Comment by probonopd** [comment]
> i'd recommend padding the runtime to a multiple of 512bytes, it will be easier to convert an appimage to a mountable disk image and back.

How so?

**Comment by harrytuttle** [comment]
By overwriting the runtime with a dos master boot record + partition table, good old fdisk will do the trick, then gnome-disk-image-mounter and other tools will detect the partition table and treat the file as a disk image. The only requirement is for the squashfs image to begin at a 512byte boundary, because dos partition tables are specified in 512bytes sectors.

Yesterday i padded the runtime-x86_64 with:
dd if=/dev/zero of=runtime-x86_64 bs=512 count=1 seek=227

Then i built an appimage with that runtime, it worked ok as an appimage. I then used fdisk to write a new partition table, with a single primary partition starting at the proper sector, then gnome-disk-image-mounter could open it with no trouble at all.

Fdisk writes zeros to the the first 446 bytes of the runtime (the partition table is from byte 447 to 512), so the runtime wasn't working anymore.

Sadly by restoring the first 446 bytes of the runtime with dd conv=notrunc bs=1 size=446 the new runtime still wouldn't work. I got the same elf format error for the old i386 runtime. But the old x86_64 runtime works!

I'd really like to find a way to insert a somewhat valid partition table into an elf: then appimages could be both executed and mounted without specifying an offset. But it's beyond my current abilities: the elf format has some hard to move structures in those 64bytes, it's not as easy as telling the linker to add a section, i tried to make some room but it didn't work. But if it works for some elf maybe there's a way. There must be hope if even a 45bytes "elf" can be abused into running:
http://www.muppetlabs.com/~breadbox/software/tiny/teensy.html
